src/internal_events/tcp.rs

```diff
@@ -190,3 +190,70 @@ impl InternalEvent for TcpBytesReceived {
         );
     }
 }
+
+pub struct TcpListenerError<'a, E> {
+    pub error: E,
+    pub message: &'a str,
+}
+
+impl<E: std::fmt::Display> InternalEvent for TcpListenerError<'_, E> {
+    fn emit(self) {
+        error!(
+            message = %self.message,
+            error = %self.error,
+            error_type = error_type::CONFIGURATION_FAILED,
+            stage = error_stage::RECEIVING,
+            internal_log_rate_secs = 10,
+        );
+    }
+}
+
+pub struct TcpSocketReceiveError<E> {
+    pub error: E,
+}
+
+impl<E: std::fmt::Display> InternalEvent for TcpSocketReceiveError<E> {
+    fn emit(self) {
+        error!(
+            message = "TCP socket error.",
+            error = %self.error,
+            error_code = "socket_failed",
+            error_type = error_type::READER_FAILED,
+            stage = error_stage::RECEIVING,
+            internal_log_rate_secs = 10,
+        );
+        counter!(
+            "component_errors_total", 1,
+            "error_code" => "socket_failed",
+            "error_type" => error_type::READER_FAILED,
+            "stage" => error_stage::RECEIVING,
+            "mode" => "tcp",
+        );
+        // deprecated
+        counter!(
+            "connection_errors_total", 1,
+            "mode" => "tcp",
+        );
+    }
+}
+
+pub struct TcpDecoderFramingError<E> {
```

Comment:
If this is a framing error, it looks identical to `DecoderFramingError` . I'd vote to just alter that struct to have the `E: std::fmt::Display` , and make use of the existing struct here.

Reply:
Yeah, I meant to comment that it was _seemingly_ the same. I didn't feel 100% confident changing the shared struct and trusting I was right 😅

Reply:
Done in https://github.com/vectordotdev/vector/pull/14422/commits/2bfdd251da4cc2eaf9b66e7fcfda9782fb47545a